### PR TITLE
Homework #5

### PR DESCRIPTION
**Pull request by ykshon527**

### CODE REVIEW

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have successfully created a state change tracking system for NBA players, which is the main requirement of the assignment. However, there are a few areas where your solution could be improved.

1) **Completeness of the solution:** Your solution does not seem to handle all the state changes as required by the prompt. Specifically, the case where a player leaves the league and then returns is not handled correctly. In your current implementation, a player who leaves the league and then returns would be marked as 'New', which is not accurate. You should modify your CASE statement to handle this scenario correctly. 

2) **Readability of the code:** Your code is generally well-structured and easy to read. You have used meaningful variable names and included comments to explain your thought process, which is good practice. However, the comments could be more detailed, especially for complex sections of the code. For example, you could explain the logic behind your CASE statement in more detail.

3) **Robustness and error handling:** Your code does not seem to handle edge cases or invalid inputs. For example, what happens if the `current_season` or `player_name` is NULL? You should include error handling to ensure your code can handle these scenarios without crashing.

4) **Modularity and reusability:** Your code is modular and could be easily adapted for other similar tasks. However, you could improve the reusability of your code by defining some of the logic (such as the state change tracking) as a separate function. This would allow you to reuse this logic in other parts of your code without having to rewrite it.

In conclusion, your solution is a good start, but there are a few areas where it could be improved. Keep up the good work and continue to refine your SQL skills!

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used window functions to calculate the number of games won by each team in a 90-game stretch. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution does not fully cover the homework prompt. The prompt asks for the most games a single team has won in a given 90-game stretch. While you have calculated the number of games won by each team in a 90-game stretch, you have not identified the team that has won the most games. You can do this by ordering the results of your final query by `max_wins` in descending order and limiting the results to the top 1.

2) **Readability:** Your code is generally well-formatted and easy to read. However, you could improve the readability of your code by adding comments to explain your thought process and the purpose of each part of your code. This will make it easier for others to understand your solution.

3) **Robustness and error handling:** Your solution does not include any error handling. While this may not be necessary for this specific task, it is a good practice to include error handling in your code to handle potential issues that may arise.

4) **Modularity and reusability:** Your solution is modular and the use of CTEs (Common Table Expressions) makes it easy to understand and reuse specific components of your code. However, you could improve the reusability of your code by parameterizing the number of games in the stretch (90 in this case). This would allow you to easily change the number of games in the stretch without having to modify multiple parts of your code.

Here is a suggestion for improving your solution:

```sql
--"What is the most games a single team has won in a given 90-game stretch?"

WITH
 nba_game_details_dedup AS (
   SELECT game_id,
     team_id,
     team_abbreviation
   from bootcamp.nba_game_details_dedup
   group by 
   game_id, 
   team_id, 
   team_abbreviation
 ),
combined AS (
  SELECT
    g.game_date_est,
    gd.team_id,
    gd.team_abbreviation,
    CASE
      WHEN gd.team_id = home_team_id
      AND home_team_wins = 1 THEN 1
      WHEN gd.team_id = visitor_team_id
      AND home_team_wins = 0 THEN 1
      ELSE 0
    END AS team_wins
  FROM
    bootcamp.nba_games g
    JOIN nba_game_details_dedup gd ON g.game_id = gd.game_id),
    
 wins_over_90 AS (
   SELECT team_id,
     team_abbreviation,
     SUM(team_wins) OVER (PARTITION BY team_id ORDER BY game_date_est ROWS BETWEEN 89 PRECEDING AND CURRENT ROW
    ) AS wins
    FROM combined
  )
    
 SELECT team_id,
   team_abbreviation,
   MAX(wins) as max_wins
    FROM wins_over_90
    ORDER BY max_wins DESC
    LIMIT 1
```

This modification will return the team that has won the most games in a 90-game stretch, which is what the homework prompt is asking for.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

The student's solution appears to cover all aspects of the homework prompt. The student has correctly used window functions to calculate the number of games won by each team over a 90-game stretch. The final query returns the team with the maximum number of wins over such a stretch, which is the expected output.

The student's solution logically follows the given homework prompt. The student first deduplicates the game details, then combines this with the games data to calculate whether each team won each game. The student then uses a window function to calculate the number of wins over a 90-game stretch for each team, and finally selects the team with the maximum number of wins.

The readability of the student's code is generally good. The student has used meaningful variable names and has structured the code in a way that is easy to follow. However, the student could improve the readability of their code by adding comments explaining their thought process and the purpose of each subquery. This would make it easier for others to understand their code.

The student's code does not include any explicit error handling. However, the nature of the task and the SQL language means that there are few potential errors that could occur. The student's code should handle edge cases and invalid inputs well, as these would likely result in SQL errors that would be easy to debug.

The student's code is modular and reusable. The student has used subqueries to break down the problem into smaller, more manageable parts. This makes it easier to understand the code and to reuse specific components in other contexts. However, the student could improve the reusability of their code by defining the subqueries as views or temporary tables, which could then be used in other queries.

Overall, the student's solution is a good attempt at the task. With a few minor improvements, particularly in terms of commenting and reusability, the student's code could be excellent.

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

**Review by prajapatipreeti — APPROVED** [review]
